### DISCUSSION TITLE
Reading any compound type to memory

### DISCUSSION THREAD

**Discussion by belugum** [BlueBrain/HighFive#820] [Questions]
Hi,

Similarly to #452, I'm trying to read a compound type unknown at compile time to memory and handle it dynamically.
I can successfully find information about the compound type using `HighFive::CompoundType::getMembers()`. But I can't read the data. 
With the following code:
```c++
std::vector<uint8_t> v;

v.resize(dataset.getStorageSize());
dataset.read(v.data());
```
I get this error:
```
HDF5-DIAG: Error detected in HDF5 (1.10.7) thread 1:
  #000: ../../../src/H5Dio.c line 198 in H5Dread(): can't read data
    major: Dataset
    minor: Read failed
  #001: ../../../src/H5Dio.c line 463 in H5D__read(): unable to set up type info
    major: Dataset
    minor: Unable to initialize object
  #002: ../../../src/H5Dio.c line 977 in H5D__typeinfo_init(): unable to convert between src and dest datatype
    major: Dataset
    minor: Feature is unsupported
  #003: ../../../src/H5T.c line 4680 in H5T_path_find(): can't find datatype conversion path
    major: Datatype
    minor: Can't get value
  #004: ../../../src/H5T.c line 4893 in H5T__path_find_real(): no appropriate function for conversion path
    major: Datatype
    minor: Unable to initialize object
terminate called after throwing an instance of 'HighFive::DataSetException'
  what():  Error during HDF5 Read. 
```
I tried other methods that either didn't compile or had similar errors.

I found that the `H5::DataSet::read` method from the HDF5 C++ API only needs a void* and the DataType, that would suit my need. Is there any equivalent in HighFive? Or another method to read a compound type dynamically (without registering a type)?

Thanks!

**Comment by 1uc** [comment]
If you have a `HighFive::DataType`, you could probably use `dset.read(ptr, datatype)`:
https://bluebrain.github.io/HighFive/class_high_five_1_1_slice_traits.html#a308151100e49985c077608619302c6c0

Looking at the code, I think it might work, depending on whether `void` is considered to be const. See,
https://github.com/BlueBrain/HighFive/blob/3addda4f3e5e91e1e8e9730829ce346ce38a8741/include/highfive/bits/H5Slice_traits_misc.hpp#L216-L234

The other semi-pragmatic answer is that what you're asking for is very close to the HDF5 C API. Maybe just use it directly? You can get the required HID from HighFive objects by calling `.getId()`.

Please let me know if these don't work, or if anything was left uncovered.
